vernac/record.ml

```diff
@@ -519,68 +598,103 @@ let implicits_of_context ctx =
 
-let declare_class def cumulative ubinders univs id idbuild paramimpls params univ arity
-    template fieldimpls fields ?(kind=Decls.StructureComponent) coers =
-  let fieldimpls =
+let build_class_constant ~univs ~rdata field implfs params paramimpls coers binder id proj_name =
+  let class_body = it_mkLambda_or_LetIn field params in
+  let class_type = it_mkProd_or_LetIn rdata.DataR.arity params in
+  let class_entry =
+    Declare.definition_entry ~types:class_type ~univs class_body in
+  let cst = Declare.declare_constant ~name:id
+      (Declare.DefinitionEntry class_entry) ~kind:Decls.(IsDefinition Definition)
+  in
+  let inst, univs = match univs with
+    | Polymorphic_entry (_, uctx) -> Univ.UContext.instance uctx, univs
+    | Monomorphic_entry _ -> Univ.Instance.empty, Monomorphic_entry Univ.ContextSet.empty
+  in
+  let cstu = (cst, inst) in
+  let inst_type = appvectc (mkConstU cstu)
+      (Termops.rel_vect 0 (List.length params)) in
+  let proj_type =
+    it_mkProd_or_LetIn (mkProd(binder, inst_type, lift 1 field)) params in
+  let proj_body =
+    it_mkLambda_or_LetIn (mkLambda (binder, inst_type, mkRel 1)) params in
+  let proj_entry = Declare.definition_entry ~types:proj_type ~univs proj_body in
+  let proj_cst = Declare.declare_constant ~name:proj_name
+      (Declare.DefinitionEntry proj_entry) ~kind:Decls.(IsDefinition Definition)
+  in
+  let cref = GlobRef.ConstRef cst in
+  Impargs.declare_manual_implicits false cref paramimpls;
+  Impargs.declare_manual_implicits false (GlobRef.ConstRef proj_cst) (List.hd implfs);
+  Classes.set_typeclass_transparency (EvalConstRef cst) false false;
+  let sub = List.hd coers in
+  let m = {
+    meth_name = Name proj_name;
+    meth_info = sub;
+    meth_const = Some proj_cst;
+  } in
+  [cref, [m]]
+
+let build_record_constant ~rdata ~ubind ~univs ~cumulative ~template fields params paramimpls coers id idbuild binder_name =
+  let record_data =
+    { Data.id
+    ; idbuild
+    ; is_coercion = false
+    ; coers = List.map (fun _ -> { pf_subclass = false ; pf_canonical = true }) fields
+    ; rdata
+    } in
+  let inds = declare_structure ~cumulative Declarations.BiFinite ~ubind ~univs paramimpls
+      params template ~kind:Decls.Method ~name:[|binder_name|] [record_data]
+  in
+  let map ind =
+    let map decl b y = {
+      meth_name = RelDecl.get_name decl;
+      meth_info = b;
+      meth_const = y;
+    } in
+    let l = List.map3 map (List.rev fields) coers (Recordops.lookup_projections ind) in
+    GlobRef.IndRef ind, l
+  in
+  List.map map inds
+
+(** [declare_class] will prepare and declare a [Class]. This is done in
+   2 steps:
+
+  1. two markely different paths are followed depending on whether the
+   class declaration refers to a constant "definitional classes" or to
+   a record, that is to say:
+
+      Class foo := bar : T.
+
+    which is equivalent to
+
+      Definition foo := T.
+      Definition bar (x:T) : foo := x.
```

Comment:
Sorry, I was mistaken, it should be 
```suggestion
      Definition bar (x:foo) : T := x.
```

Reply:
I should have checked sorry, lazy me :/ Will amend in a second, thanks @SkySkimmer